lasagne/updates.py

```diff
@@ -567,6 +575,48 @@ def adam(loss_or_grads, params, learning_rate=0.001, beta1=0.9,
     return updates
 
 
+def step_scaling(tensor_vars, threshold):
+    """Rescales a list of tensors based on their combined norm
+
+    If the combined norm of the input tensors exceeds the threshold then all
+    tensors are rescaled such that the combined norm is equal to the threshold.
+
+    Parameters
+    ----------
+    tensor_vars: TensorVariable or list of TensorVariables
+        Tensors to be rescaled.
+    threshold : float
+        Threshold value for total norm.
+
+    Returns
+    -------
+    List of:
+
+        * List of scaled tensors
+        * Total norm
+        * Scaling multiplier
+
+    Notes
+    -----
+    The total norm and the scaling multiplier can be used to monitor training.
+
+    """
+    if not isinstance(tensor_vars, (list, tuple)):
+        tensor_vars = [tensor_vars]
+
+    sq = 0
+    for tensor in tensor_vars:
+        sq += (tensor**2).sum()
+
+    norm = T.sqrt(sq)
+    threshold = theano.shared(utils.floatX(threshold))
+    scale = threshold / norm
+    multiplier = T.switch(norm < threshold, 1.0, scale)
+    tensor_vars_scaled = [step*multiplier for step in tensor_vars]
```

Comment:
I'm wondering if this is better or worse than what we have in `norm_constraint`. It will fail for a norm of zero (that's bad), and it uses `T.switch` instead of `T.clip` for the case distinction (not sure which one is faster).